[Dependency Scanning] Bridge Clang dependency scanner results on-demand

artemcm opened the pull request:
https://github.com/swiftlang/swift/pull/83050 changed the behavior to query every Clang module identifier imported from Swift, even for modules which may have previously been discovered as transitive dependencies of another query. This means we now have a lot of results coming in from Clang scanner queries which are redundant, and we bridge them all, all of the time, which adds up to a substantial amount of time.

Instead of always bridging all of the discovered modules of all of the queries, only do so for modules which have not already been cached at the time we record query results.

artemcm committed:
[Dependency Scanning] Bridge Clang dependency scanner results on-demand

Instead of always bridging all of the discovered modules of all of the queries, only do so for modules which are not already cached

artemcm commented:
@swift-ci smoke test

cachemeifyoucan reviewed (commented):
Function wise it is all good. Just some comments about the APIs.

artemcm committed:
[Dependency Scanning] Refactor Clang dependency bridging into a 'ModuleDependencyScanner' utility

This moves the functionality of 'bridgeClangModuleDependency' into a utility in the main scanner class because it relies on various objects whose lifetime is already tied to the scanner itself.

artemcm commented:
@swift-ci smoke test

artemcm commented on include/swift/AST/ModuleDependencies.h:
```diff
@@ -174,6 +176,18 @@ void registerCxxInteropLibraries(
+                                   clang::tooling::dependencies::ModuleOutputKind)>;
+
+ModuleDependencyInfo
+bridgeClangModuleDependency(
```
This required a bigger refactor than it originally seemed, but I moved the bridging code into a private scanner utility.

artemcm commented:
@swift-ci smoke test

artemcm commented:
@swift-ci smoke test Linux platform

cachemeifyoucan reviewed (commented):
Mostly better. Still some comments about how functions are layered.

artemcm commented:
@swift-ci smoke test

artemcm commented on include/swift/DependencyScan/ModuleDependencyScanner.h:
```diff
@@ -40,41 +40,26 @@ class ModuleDependencyScanningWorker {
-  /// \returns \c true if an error occurred, \c false otherwise
-  bool scanHeaderDependenciesOfSwiftModule(
-      const ASTContext &ctx,
+  /// of a Swift module.
```
Fixed.

artemcm commented:
@swift-ci smoke test
